modules/dlm/src/main/java/org/elasticsearch/dlm/DataLifecycleService.java

```diff
@@ -331,6 +441,84 @@ public void onFailure(Exception e) {
         });
     }
 
+    /*
+     * This method executes the given force merge request. Once the request has completed successfully it writes a timestamp as custom
+     * metadata in the cluster state indicating when the force merge has completed. The listener is notified after the cluster state
+     * update has been made, or when the forcemerge fails or the write of the to the cluster state fails.
+     */
+    private void forceMergeIndex(ForceMergeRequest forceMergeRequest, ActionListener<Void> listener) {
+        assert forceMergeRequest.indices() != null && forceMergeRequest.indices().length == 1 : "DLM force merges one index at a time";
+        final String targetIndex = forceMergeRequest.indices()[0];
+        logger.info("DLM is issuing a request to force merge index [{}]", targetIndex);
+        client.admin().indices().forceMerge(forceMergeRequest, new ActionListener<>() {
+            @Override
+            public void onResponse(ForceMergeResponse forceMergeResponse) {
+                if (RestStatus.OK.equals(forceMergeResponse.getStatus())) {
+                    if (forceMergeResponse.getFailedShards() > 0) {
+                        String message = Strings.format(
+                            "Force merge request resulted in %d failed shards",
+                            forceMergeResponse.getFailedShards()
+                        );
+                        listener.onFailure(new ElasticsearchException(message));
+                        logger.warn(message);
```

Comment:
nit: This might be a personal preference, but should we log the message before we notify the listeners? (to avoid the cognitive context switch of "leaving this method" )

I think this could be simplified by doing something like:

```
            if (response.getFailedShards() == 0) {
                      logger.info("DLM successfully force merged index [{}]", targetIndex);
                      setForceMergeCompletedTimestamp(targetIndex, listener);
           } else {
                // log a warn message that has the shard failures
                DefaultShardOperationFailedException[] failures = response.getShardFailures();
String errorMessage = String.format(
                    Locale.ROOT,
                    "DLM failed to forcemerge all shards for inded [%s] due to failures [%s]",
                    indexName,
                    failures == null
                        ? "n/a"
                        : Strings.collectionToDelimitedString(
                            Arrays.stream(failures).map(Strings::toString).collect(Collectors.toList()),
                            ","
                        )
                );
                logger.warn(errorMessage);
                // report failure and we'll retry next time DLM runs
                listener.onFailure(new ElasticsearchException(errorMessage));
           }
```

Or something along those lines? What do you think?

Reply:
I think I can get rid of the status check though, since getStatus() is just a check of the number of failed shards:
```
    public RestStatus getStatus() {
        if (failedShards > 0) {
            return shardFailures[0].status();
        } else {
            return RestStatus.OK;
        }
    }
```

Reply:
I'm pretty sure this invariant holds 

> Or if you can show me that we're guaranteed that getTotalShards() == getSuccessfulShards() if getFailedShards() == 0 then I'd be happy to merge the two blocks.

Reply:
It seems odd that the response has total, successful, and failed if total is always equal to successful + failed. That was my reasoning behind separating that out. I definitely didn't go into the lucene code to see what really happens.

Reply:
Coincidentally, ExplainLifecycleDataIT.testExplainLifecycleForIndicesWithErrors() just exposed a case where the number of failed shards is 0, but the total number of shards was not equal to the number of successful shards. I think in that case it is because shards are being allocated. I think in this case we do want it to fail so that we run forcemerge again next time to merge the segments in the shards we missed.